Core/Src/main.c

```diff
@@ -1,181 +1,252 @@
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+const double WARG_MAX_DUTY_CYCLE = 0.1;
+const double WARG_MIN_DUTY_CYCLE = 0.05;
+const uint32_t WARG_ADC_MAX_VALUE = 1023;
+const uint32_t WARG_TIM1_COUNTER_PERIOD = 64000;
+const uint32_t WARG_TIM1_PRESCALER = 14;
+const uint32_t WARG_TIM1_PWM_CHANNEL = 1;
+const uint32_t WARG_ADC_SELECT = 0b1000;
+const uint32_t WARG_GPIO_PIN = 8;
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  // Initialize SPI1
+  SPI_HandleTypeDef hspi1;
+  HAL_SPI_Init(&hspi1);
+  hspi1.Init.Mode = SPI_MODE_MASTER;
+  hspi1.Init.Direction = SPI_DIRECTION_2LINES;
+  hspi1.Init.DataSize = SPI_DATASIZE_8BIT;
+  hspi1.Init.CLKPolarity = SPI_POLARITY_LOW;
+  hspi1.Init.CLKPhase = SPI_PHASE_1EDGE;
+  hspi1.Init.NSS = SPI_NSS_SOFT;
+  hspi1.Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_16;
+  hspi1.Init.FirstBit = SPI_FIRSTBIT_MSB;
+  hspi1.Init.TIMode = SPI_TIMODE_DISABLE;
+  hspi1.Init.CRCCalculation = SPI_CRCCALCULATION_DISABLE;
+
+  //Initialize TIM1
+  TIM_HandleTypeDef htim1;
+  HAL_TIM_PWM_Init(&htim1);
+  htim1.Init.Prescaler = WARG_TIM1_PRESCALER;
+  htim1.Init.CounterMode = TIM_COUNTERMODE_UP;
+  htim1.Init.Period = WARG_TIM1_COUNTER_PERIOD;
+  htim1.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1; // No division?
+  htim1.Init.RepetitionCounter = 0;
+  TIM_OC_InitTypeDef octim1;
+  octim1.OCMode = TIM_OCMODE_PWM1;
+  octim1.Pulse = 0;
+  octim1.OCPolarity = TIM_OCPOLARITY_HIGH;
+  octim1.OCNPolarity = TIM_OCNPOLARITY_HIGH;
+  octim1.OCFastMode = TIM_OCFAST_DISABLE;
+  octim1.OCIdleState = TIM_OCIDLESTATE_RESET;
+  HAL_TIM_PWM_ConfigChannel(&htim1, &octim1, WARG_TIM1_PWM_CHANNEL);
+
+  uint16_t size = 2;
+  uint8_t pTxData[2] = {WARG_ADC_SELECT, 0}; // select ch0, single for ADC
+  uint8_t pRxData[2];
+  uint32_t timeout = 0;

```

Comment:
Couple things here: 

1) Same as `size` I noticed that `timeout` does not change, should it be const for readability?

2) I see that you set your timeout as 0. This would be an issue when you use it in `HAL_SPI_TransmitReceive` because by making the timeout 0 you literally give no time for the peripheral to get a response so it will always return with an error code.

Reply:
Fixed now